File: WorkflowPocBackend/WorkflowPocBackend.API/CamundaService.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Text;
using Newtonsoft.Json;

namespace WorkflowPocBackend
{
	internal class CamundaService
	{
		private const string baseUrl = "http://localhost:8080/engine-rest/";
		private const string processDefinitionKey = "onboarding";
		private const string selectOfferingTopic = "selectOffering";
		private const string captureOfferingDetailsTopic = "captureOfferingDetails";
		private const string addUserTopic = "addUser";
		private const string initialiseOfferingTopic = "initialiseOffering";
		private const string sendWelcomeLetterTopic = "sendWelcomeLetter";

		private static readonly HttpClient httpClient = new HttpClient(new HttpClientHandler
		{
			Proxy = new WebProxy { BypassProxyOnLocal = true }
		});

		internal void StartInstance()
		{
			var payload = new StringContent(JsonConvert.SerializeObject(new { processDefinitionKey }), Encoding.UTF8, "application/json");
			var result = httpClient.PostAsync($"{baseUrl}process-instance", payload).Result;
			var resultString = result.Content.ReadAsStringAsync().Result;
			var instances = JsonConvert.DeserializeObject<ProcessInstanceDto[]>(resultString);
			foreach (var instance in instances)
			{
				Console.WriteLine($"Deleting instance {instance.id}");
				result = httpClient.DeleteAsync($"{baseUrl}process-instance/{instance.id}").Result;
				if (!result.IsSuccessStatusCode)
					throw new Exception($"{result.StatusCode}");
			}

			payload = new StringContent("{}", Encoding.UTF8, "application/json");
			result = httpClient.PostAsync($"{baseUrl}process-definition/key/{processDefinitionKey}/start", payload).Result;
			if (!result.IsSuccessStatusCode)
				throw new Exception($"{result.StatusCode}");

			Console.WriteLine("Started instance");
		}

		internal void SelectOffering(string offering)
		{
			CompleteTask(selectOfferingTopic, new { offering = new { value = offering } });
		}

		internal void CaptureOfferingDetails()
		{
			CompleteTask(captureOfferingDetailsTopic);
		}

		internal void AddUser(bool addAnother)
		{
			CompleteTask(addUserTopic, new { addAnother = new { value = addAnother } });
		}

		internal void InitialiseOffering()
		{
			CompleteTask(initialiseOfferingTopic);
		}

		internal void SendWelcomeLetter()
		{
			CompleteTask(sendWelcomeLetterTopic);
		}

		internal string GetOfferingType()
		{
			var resultString = Call($"{baseUrl}variable-instance", new { variableName = "offering" });
			var variables = JsonConvert.DeserializeObject<List<VariableDto>>(resultString);
			return variables[0].value;
		}

		private void CompleteTask(string topic, object variables = null)
		{
			var workerId = Guid.NewGuid().ToString();

			var fetchAndLockPayload = new
			{
				maxTasks = 1,
				workerId,
				topics = new[]
				{
					new
					{
						topicName = topic,
						lockDuration = 20000
					}
				}
			};

			var resultString = Call($"{baseUrl}external-task/fetchAndLock", fetchAndLockPayload);

			var tasks = JsonConvert.DeserializeObject<List<TaskInstanceDto>>(resultString);
			if (tasks.Count == 0)
			{
				Console.WriteLine($"No {topic} task found.");
				return;
			}

			Console.WriteLine($"{topic} task variables: {JsonConvert.SerializeObject(tasks[0].variables)}");

			var completePayload = new
			{
				workerId,
				variables = variables ?? new object()
			};

			Call($"{baseUrl}external-task/{tasks[0].id}/complete", completePayload);

			Console.WriteLine($"{topic} task completed.");
		}

		private string Call(string url, object requestData)
		{
			var payload = new StringContent(JsonConvert.SerializeObject(requestData), Encoding.UTF8, "application/json");
			var result = httpClient.PostAsync(url, payload).Result;
			return result.Content.ReadAsStringAsync().Result;
		}

		private class ProcessInstanceDto
		{
			public string id;
		}

		private class TaskInstanceDto
		{
			public string id;
			public object variables;
		}

		private class VariableDto
		{
			public string value;
		}
	}
}

File: src/app/camunda.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";

@Injectable()
export class CamundaService {

    /*






                    This service didn't end up being used - it contains methods to call Camunda directly, 
                    and requires the proxy.conf.json file referenced in angular.json to avoid CORS issues







    */


    private readonly baseUrl = 'http://localhost:4200/engine-rest/';
    private readonly processDefinitionKey = 'onboarding';

    constructor(private httpClient: HttpClient) { }

    startOnboarding(onSuccess: () => void): void {
        this.getAndDeleteInstances(() => {
            console.log(`Starting new instance`);
            this.httpClient.post(`${this.baseUrl}process-definition/key/${this.processDefinitionKey}/start`, {})
                .subscribe(() => {
                    console.log('Started instance');
                    onSuccess();
                });
        });
    }

    selectOffering(offering: string, onSuccess: () => void): void {
        const workerId = "offeringWorkerId"; // TODO: In real project, workerId should be generated for each fetch-complete cycle
        const topic = "selectOffering";

        this.fetchTask(workerId, topic, task => {
            this.completeTask(workerId, { "offering": { "value": offering } }, task.id, onSuccess);
        });
    }

    private fetchTask(workerId: string, topic: string, onSuccess: (task: TaskInstance) => void) {
        const body = {
            "workerId": workerId,
            "maxTasks": 1,
            "topics": [
                {
                    "topicName": topic,
                    "lockDuration": "10000"
                }
            ]
        };
        this.httpClient.post<TaskInstance[]>(`${this.baseUrl}external-task/fetchAndLock`, body)
            .subscribe(tasks => {
                console.log(`Fetched task ${tasks[0].id} for topic ${topic}`);
                onSuccess(tasks[0]);
            });
    }

    private completeTask(workerId: string, variables: Object, taskId: string, onSuccess: () => void) {
        const body = {
            "workerId": workerId,
            "variables": variables
        };
        this.httpClient.post(`${this.baseUrl}external-task/${taskId}/complete`, body)
            .subscribe(() => {
                console.log(`Completed task ${taskId}`);
                onSuccess();
            });
    }

    private getAndDeleteInstances(onSuccess: () => void): void {
        console.log('Getting instances');

        this.httpClient.post<ProcessInstance[]>(`${this.baseUrl}process-instance`, { 'processDefinitionKey': this.processDefinitionKey })
            .subscribe(instances => {
                console.log(instances);
                console.log('Deleting instances');
                this.deleteInstances(instances, onSuccess);
            });
    }

    private deleteInstances(instances: ProcessInstance[], onSuccess: () => void): void {
        if (instances.length > 0) {
            console.log(`Deleting instance ${instances[0].id}`);
            this.httpClient.delete(`${this.baseUrl}process-instance/${instances[0].id}`)
                .subscribe(() => { this.deleteInstances(instances.splice(1), onSuccess); });
        } else {
            onSuccess();
        }
    };
}

interface ProcessInstance {
    id: string;
}

interface TaskInstance {
    id: string;
}